src/ast/irbuilderbpf.cpp

```diff
@@ -498,10 +505,14 @@ Value *IRBuilderBPF::createScratchBuffer(
   auto cmp = CreateICmp(CmpInst::ICMP_ULE, cpu_id, max, "cpuid.min.cmp");
   auto select = CreateSelect(cmp, cpu_id, max, "cpuid.min.select");
 
+  // Note the 1st index is 0 because we're pointing to
+  // ValueType var[MAX_CPU_ID + 1][num_elements]
+  // 2nd/3rd/4th indexes actually index into the array
+  // See https://llvm.org/docs/LangRef.html#id236
   return CreateGEP(GetType(type),
                    CreatePointerCast(module_.getGlobalVariable(config.name),
                                      GetType(type)->getPointerTo()),
-                   { select, getInt64(key), getInt64(0) });
+                   { getInt64(0), select, getInt64(key), getInt64(0) });
```

Comment:
TL;DR: There was a bug in my createScratchBuffer() function but my analysis indicates it doesn't cause bad effects for fmtstr as it only affects where key > 0. So I think a forward fix is fine. But let me know if I'm wrong.

@danobi @jordalgo this was a bug in my initial fmtstrings implementation. Basically because the type was a pointer to an array, the first index needs to be 0 to get the actual array itself. See https://llvm.org/docs/LangRef.html#id236

In theory, this should cause out-of-bounds access in my previous PR for CPU IDs > 0

The only reason I don't think it did is because the LLVM IR instruction:
```
%12 = getelementptr [1 x [1 x [24 x i8]]], ptr @fmt_str_buf, i64 %cpuid.min.select, i64 0, i64 0
```

Generates something like:
```
r9 = <global var address>
16: (85) call bpf_get_smp_processor_id#8
17: (2d) if r6 > r0 goto pc+1         ; R0_w=55 R6=55
18: (bf) r0 = r6                      ; R0=55 R6=55
19: (27) r0 *= 48                     ; R0_w=2640
20: (0f) r9 += r0
```

Instruction 19 is basically multiplying bpf_get_smp_processor_id() * sizeof([1 x [1 x [24 x i8]]]) which is 24.

So even with multiple CPUs involved, the BPF assembly still works because the real data structure is [N x [1 x [24 x i8]]] as key is always 0 in fmt strings buffer.

But for this PR where key >= 0, without this fix, `getInt64(key)` thinks it indexing into `[1 x [2 x [[24 x i8]]])` and not `[2 x [[24 x i8]]` so we end up overrunning the buffer. For example, I saw stuff like:
```
r1 = 6
*(u64 *)(r9 +56) = r1
```

Reply:
Nice catch! GEP is tricky like that :P

Reply:
Sounds good, I'll split this into 2 commits